### ISSUE TITLE
explore interoperability with other agnostic-wrap libraries

### ISSUE THREAD

**Issue by lucascolley** [mdhaber/marray#67]
Once https://github.com/lucascolley/pint-array is relatively stable we should take a look at wrapping arrays from each library with the other library. Ideally the behaviour should be the same for both ways.

**Comment by vnmabus** [comment]
To be honest, I would love to have a general "array with metadata" library. That is, a library that:

1. Provides a type `MetadataArray` which wraps a standard array API object, and exposes the wrapped array object as some attribute with a standard name (or provides a function to retrieve it).
2. Allows to attach metadata with different keys (e.g. mask, units, row_names, column_names, etc). The metadata itself would be array-likes broadcastable with the wrapped array object, and conform to a `Metadata` protocol that provide methods/functions to specify the behavior when operating arrays with those keys. For example, we could request that the `Metadata` arrays have a method with the following signature:
````python
def compute_metadata(key: str, xp_namespace: Mapping[str, Callable], function_name: str, *args: Any, **kwargs: Any) -> Metadata | MetadataArray | None | NotImplemented
````
When executing a function such as `xp.add(x1, x2)`, the `MetadataArray` would check all keys present in the involved arrays, and call that method for the first `Metadata` object present with each key, as `metadata.compute_metadata(<key>, xp, "add", x1, x2)`. The method could:
- Return the `Metadata` object corresponding to the key in the result.
- Return the full resulting `MetadataArray`, for metadata that needs to affect the computations of the wrapped array (such as mask in the case of reductions). It is not yet clear to me if it is possible to somehow support several keys trying to do this.
- Raise an exception, to abort the operation.
- Return `None`, to clear the metadata key in the result.
- Return `NotImplemented`, so that the next `Metadata` object with the same key has a chance to run its method.

I think that with this providing the common infrastructure, the mask, units or label support can be done more easily by subclassing `Metadata` and defining the `compute_metadata` method. Moreover, we would reduce the need to stack a lot of array wrappers, which is problematic in order to know if some layer provides a particular metadata attribute (it depends on the order in the stack).

@lucascolley, @mdhaber, what do you think about this? Do you think this is a viable/desirable approach? Should I propose that somewhere else?

**Comment by lucascolley** [comment]
> Moreover, we would reduce the need to stack a lot of array wrappers, which is problematic in order to know if some layer provides a particular metadata attribute (it depends on the order in the stack).

This is definitely a big problem that we need to address 👍 thanks for thinking about this @vnmabus !

> Return the full resulting MetadataArray, for metadata that needs to affect the computations of the wrapped array (such as mask in the case of reductions). It is not yet clear to me if it is possible to somehow support several keys trying to do this.

That is my main concern. My feeling is that it may not be viable to support arbitrary combinations of wrapping libraries. It might still be worth trying though, if everything seems to work out.

---

I had been thinking along slightly different lines, where array objects would expose a namespace via a `.metadata` attribute (name TBD). Use of wrapping libraries would then mimic use of extension modules:

```python
if hasattr(xp, "linalg"):
	...
# similar to
if hasattr(x, "metadata.mask"):
	...
```

It would then be simple for wrapping libraries to 'propagate metadata up the stack' by populating their `metadata` namespace with entries from the array they are wrapping. The big problem with this approach is that we would want to modify the underlying array object for non-compliant libraries, which is [out of scope for array-api-compat](https://data-apis.org/array-api-compat/dev/special-considerations.html#special-considerations). Perhaps this isn't a problem if unwrapped arrays would always have this namespace empty, though?

**Comment by lucascolley** [comment]
> Should I propose that somewhere else?

We should definitely have a discussion about this on the array-api repo at some point, but perhaps no rush.

**Comment by vnmabus** [comment]
> That is my main concern. My feeling is that it may not be viable to support arbitrary combinations of wrapping libraries. It might still be worth trying though, if everything seems to work out.

This is my main concern too. However I am not sure if there are many other cases, apart from masking, in which we really want the metadata to affect the results of the computation.

> It would then be simple for wrapping libraries to 'propagate metadata up the stack' by populating their metadata namespace with entries from the array they are wrapping.

I also thought about that. If we choose the stack wrapper model, I would propose to have a standard name for the wrapped array, as well as a function to retrieve particular `Metadata` arrays by iterating through the stack following the chain.

> We should definitely have a discussion about this on the array-api repo at some point, but perhaps no rush.

I actually opened a (less detailed) [discussion ]( https://github.com/data-apis/array-api/issues/712)more than one year ago, but at that moment it was deemed out-of-scope. They pointed me to https://github.com/pydata/duck-array-discussion/issues, but the discussion there seems to have stalled. I would love to have a centralized place to discuss this topic, and you seemed to have the expertise to bring the discussion forward.

I was bitten by this because I am trying to extend in the long term my own types in [scikit-fda](https://fda.readthedocs.io/en/latest/) to provide general types to represent discretized or series-based arrays of functions with tensor input and output, compatible with the array API standard. However it would be good if I could attach metadata such as labels or units without having to (poorly) reimplement the wheel.

**Comment by crusaderky** [comment]
I am *extremely* excited about the idea of nested wrapping.

However, I think that the most serious issue is when you have a library, such as `array-api-extra`, that special-cases a namespace (in `xpx.at`, `xpx.lazy_apply`, or `xpx.apply_where`). At that point, even if you have a standardized way to travel into the data of the inner namespace, everything falls apart because the generic library (array-api-extra) doesn't know how to build the output _metadata_.

For example, it's impossible for `xpx.lazy_apply` to accept `pint-array` objects, either explicitly or by means of a standardized `get_wrapped_array(wrapper)` API:

```python
import pint_array
import array_api_compat.numpy
import array_api_extra

xp = pint_array.wrap_namespace(array_api_compat.numpy)

s = xp.asarray([10, 20], units="km")
t = xp.asarray([3, 4], units="s")

# Stand-ins for Cython or Numba functions
kernel1 = lambda s, t: s / t
kernel2 = lambda s, t: s * t

out1 = xpx.lazy_apply(kernel1, s, t, as_numpy=True)  # out1.units = km / s
out2 = xpx.lazy_apply(kernel2, s, t, as_numpy=True)  # out2.units = km * s
```
In the above example, `array_api_extra` does not have the information to generate the two output `units`.
It already has this problem with Dask, where it is impossible to apply it to chunked inputs because it doesn't know what the function will do with it.

`lazy_apply` could recursively call itself until it reaches a non-wrapper array - but it still can't solve this problem.

MArray has the same problem. Consider feeding these to `xpx.lazy_apply`:
```python
kernel1 = lambda m, a, b: mxp.where(m, a, b)  # out.mask = m.mask | (where(m.data, a.mask, b.mask)
kernel2 = lambda m, a, b: a + mxp.where(m, 2, 3) * b  # out.mask = m.mask | a.mask | b.mask
```

mxp above is another can of worms, but at least it's _solvable_: https://github.com/data-apis/array-api-extra/issues/160

**Comment by lucascolley** [comment]
> Cython or Numba functions ... array_api_extra does not have the information to generate the two output units.

Sure, I think we would have to specify the unit behaviour separately (same for mask behaviour) when things leave python-land, then provide this information to `xpx.lazy_apply` (TBD on how to make that work with the signature in the general case). Then we can call the kernel on `get_wrapped_array(x)`, use another helper to insert the kernel result back into the wrapper, and afterwards `quantity_namespace.asarray(wrapped_kernel_result, mask=...)` where the mask would need to be uniquely determined by whatever information is provided (thinking back to previous discussion at https://github.com/scipy/scipy/pull/22639#discussion_r1985151312).

The programming of all this re-wrapping of the kernel result for arbitrarily many different wrappers is obviously something which requires more thought.

**Comment by mdhaber** [comment]
> everything falls apart because the generic library (array-api-extra) doesn't know how to build the output metadata.

Isn't the trouble because the implementations in `array-api-extra` aren't generic? 

The library is serving two purposes now: 
- collecting routines that can be expressed generically in terms of the standard but are deemed too specific for

**Comment by crusaderky** [comment]
> I'm hoping that when we see usage of things like `apply_where` and `at.set`, they can be added to the standard, and eventually the standard will be complete enough that special cases aren't needed in every-day functions.

I expect that `at` will be added to the standard eventually.
`apply_where` intrinsically depends on either boolean indexing plus being writable, or exposing the meta-namespace to the user. 
- Needing boolean indexing plus being writable would mean that the function would be allowed to fall back to `where(cond, f1(*args), f2(*args))`. This is what `xpx.apply_where` currently does on JAX, which is only a performance problem _until f1 or f2 is a `jax.pure_callback` / `lazy_apply`_ , at which point it can cause a segfault or an infinite loop!
- exposing the meta-namespace, or in other words applying f1/f2 to the innermost wrapped array (this is what `xpx.apply_where` does for Dask), incurs again in the problem of `apply_where` not knowing how to re-wrap the output with correct metadata.

> use another helper to insert the kernel result back into the wrapper, and afterwards `quantity_namespace.asarray(wrapped_kernel_result, unit=...)` where the unit would need to be uniquely determined by whatever information is provided

This is the thing; it's not a problem to agree on a standard API to unwrap/rewrap the _data_ programmatically; generating the _metadata_ for the output requires information that only the outermost caller (whoever is calling `xpx.lazy_apply(f, ...)` and knows what `f` does)  can provide.

Just consider `scipy.special` (https://github.com/scipy/scipy/pull/22639). 
There, we have explicit unwrap->apply->rewrap for marray where the output mask is an elementwise OR-reduction of the args' masks. We can do that because we know that _all_ functions in that module are elementwise. That logic would fall apart with a gufunc, e.g. `z = x.sum(axis=0) + y`.
Even with just ufuncs, though, the same unwrap->apply->rewrap pattern would be _drastically_ more complicated to pull with `pint-array`, as you would need to know the exact arithmetics that each and every function operates on - at which point I suspect you'd be better off just providing an array-agnostic variant of everything.
